src/main/java/com/powsybl/openloadflow/ac/nr/NewtonRaphson.java

```diff
@@ -63,7 +65,36 @@ public static List<Pair<Equation<AcVariableType, AcEquationType>, Double>> findL
                 .collect(Collectors.toList());
     }
 
-    private NewtonRaphsonStatus runIteration(StateVectorScaling svScaling, MutableInt iterations) {
+    public static void reportLargestMismatch(Reporter reporter, EquationSystem<AcVariableType, AcEquationType> equationSystem, double[] mismatch, LfNetwork network, int iteration) {
+        List<Pair<Equation<AcVariableType, AcEquationType>, Double>> mismatchEquations = equationSystem.getIndex().getSortedEquationsToSolve().stream()
```

Comment:
You want to find the largest mismatch by equation type, so you want to group the mismatches by equation type and retain only the max by mismatch. There's `Collectors::groupingBy` which does exactly this kind of operation. This will simplify things, as you don't need to create a full list which will be later filtered by equation type.
```java
Map<AcEquationType, Optional<Pair<Equation<AcVariableType, AcEquationType>, Double>>> mismatchEquations = equationSystem.getIndex().getSortedEquationsToSolve().stream()
        .map(equation -> Pair.of(equation, mismatch[equation.getColumn()]))
        .collect(Collectors.groupingBy(e -> e.getKey().getType(), Collectors.maxBy(Comparator.comparingDouble(e -> Math.abs(e.getValue())))));
```
I don't think it's still useful to compare with `Math.pow(10, -7)` in that case.

Reply:
What could be the purpose of having something like Map<X, Optional<Y>> ?
If optional is empty, there is no reason to have not in the map, right?
So why not just Map<X, Y> ?